react/components/GradientCollapse/index.js

```diff
@@ -10,100 +10,91 @@ const transitionStyle = transitionTime => ({
   transition: `${transitionTime}ms ease-in-out`,
 })
 
-class GradientCollapse extends Component {
-  constructor(props) {
-    super(props)
-    this.state = { isCollapseVisible: true, collapsed: true, maxHeight: 'auto' }
+const fadeBottomClasses = state =>
+  classNames(styles.fadeBottom, { 'o-0': state === 'entered' }, 'w-100 h-50')
+const pointerEventsAutoClasses = state =>
+  classNames(
+    styles.pointerEventsAuto,
+    {
+      'bg-transparent': state === 'entered',
+      'bg-base': state != 'entered',
+    },
+    'tc w-100'
+  )
 
-    this.wrapper = React.createRef()
-  }
-
-  calcMaxHeight = () => {
-    const { collapseHeight } = this.props
-    const wrapper = this.wrapper.current
+function GradientCollapse(props) {
+  const { children, collapseHeight } = props
+  const [collapsed, setCollapsed] = useState(true)
+  const [maxHeight, setMaxHeight] = useState('auto')
+  const [collapseVisible, setCollapseVisible] = useState(true)
+  const wrapper = useRef()
 
-    if (wrapper.scrollHeight > collapseHeight) {
-      const maxHeight = wrapper.scrollHeight + 60
-      this.setState({ isCollapseVisible: true, maxHeight })
-    } else this.setState({ isCollapseVisible: false, maxHeight: 'auto' })
+  const calcMaxHeight = () => {
+    const wrapperEl = wrapper.current
+    if (wrapperEl.scrollHeight > collapseHeight) {
+      setMaxHeight(wrapperEl.scrollHeight + 60)
+      setCollapseVisible(true)
+    } else {
+      setCollapseVisible(false)
+      setMaxHeight('auto')
+    }
   }
 
-  debouncedCalcMaxHeight = debounce(this.calcMaxHeight, 500)
+  const debouncedCalcMaxHeight = debounce(calcMaxHeight, 500)
+  useEffect(() => {
+    window.addEventListener('resize', debouncedCalcMaxHeight)
+    calcMaxHeight()
+    return () => {
+      window.removeEventListener('resize', debouncedCalcMaxHeight)
+    }
+  })
```

Comment:
missing useEffect deps like []

Reply:
It doesn't work if I do this, because of the function debouncedCalcMaxHeight that needs to be updated after every render

Reply:
Because of the closure of the function, I have to update this every time

Reply:
useEffect tem que passar [], é obrigatório, não passar causa efeitos estranhos.

O problema que você esta relatando é solucionado ao corretamente usar o hook useCallback.

const calcMaxHeight = useCallback(() => {... implementaçao}, [dependencias dessa funcao])
const debouncedCalcMaxHeight = useCallback(debounce(calcMaxHeight, 500), [calcMaxHeight])
useEffect(() => {...}, [debouncedCalcMaxHeight, calcMaxHeight])

Mas se você quer simular um comportamento similar ao componentDidMount, eu falho em ver como passar [] para as dependencias causa algum problema.

Reply:
Pois é, eu não quero replicar o componentDidMount (acho que ficou meio confuso pq eu passei o componente pra FC e fiz um fix na msmPR), o que eu to querendo fazer é um didMount + didUpdate.

Reply:
Mas fiz a alteração pra usar o useCallback

Reply:
Eu tinha colocado useCallback, mas vi agr que não funciona, aí tirei de novo